### PR TITLE
Change RSTE of RTRS to ReturnOnly by default

### LINKED ISSUE #64122: Change RTRS default escape to ReturnOnly
Follow-up to #64100

In `M` below, the default scope of parameter `rs` will change from `scoped` to *ReturnOnly*. It will be an error to use `[UnscopedRef]`.

```cs
void M(ref RS rs) { }
ref struct RS { public ref int field; }
```

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #64122

### CODE REVIEW

**Comment by Sergio0694** [comment]
I've been talking with @jaredpar on Discord about potentially making the default STE scope of `out` be return-only too, post C# 11 (he mentioned we don't have time to make it happen in C# 11 at this point). Do we want to also have tests to show the scenario that's blocked currently following this change, so we also have them ready to go when/if we flip this in C# 12?

We might very well already have such tests, but just for reference, one example could be:

```csharp
ref struct Foo
{
    public byte Value;

    public static Bar(ref Foo foo, out Span<byte> span)
    {
        span = new(ref foo.Value); // Error now, should be valid in C# 12
    }
}
```

So that following this, you'd also have:

```csharp
static Span<byte> Baz(scoped ref Foo foo)
{
    Foo.Bar(ref foo, out var span); // Ok

    return span; // Error
}
```

At least I'm pretty sure that's how that was supposed to be, I hope I'm not mixing things up again 😅

**Comment by jaredpar** [comment]
@Sergio0694 #64155 :smile:

**Comment by Sergio0694** [comment]
Oh wow I had missed that, well please ignore me then ahah
(but also, yay 🎉)

**Comment by cston** [comment]
>     static ref C F7(out C c) { c = default; return ref c; } // 3

Consider updating numbered comments.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:1480 in f00f7d8. [](commit_id = f00f7d83a524f9336843ca496436223e89af9897, deletion_comment = False)

**Comment by cston** [comment]
>         D5 d5 = (ref R x, ref R y) => ref x;

Consider renumbering. For instance, this should be `// 2`.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:9395 in f00f7d8. [](commit_id = f00f7d83a524f9336843ca496436223e89af9897, deletion_comment = False)

**Comment by cston** [comment]
>         D5 d5 = M3;

Consider renumbering: `// 2`, etc.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:9454 in f00f7d8. [](commit_id = f00f7d83a524f9336843ca496436223e89af9897, deletion_comment = False)

**Comment by cston** [comment]
>         D5 d5 = delegate(ref R x, ref R y) { return ref x; };

Consider renumbering: `// 2`, etc.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:9595 in f00f7d8. [](commit_id = f00f7d83a524f9336843ca496436223e89af9897, deletion_comment = False)

**Comment by cston** [comment]
>     public R<string> F4(ref R<string> r) => default;

`// 2`, etc.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:10548 in f00f7d8. [](commit_id = f00f7d83a524f9336843ca496436223e89af9897, deletion_comment = False)

**Comment by cston** [comment]
>     R<string> I<string>.F4(ref R<string> r) => default;

It looks like `I<string>.F4(ref R<string> r)` should be reported as an error.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:10586 in f00f7d8. [](commit_id = f00f7d83a524f9336843ca496436223e89af9897, deletion_comment = False)

**Comment by cston** [comment]
>         var f3 = new[] { (ref R r) => { }, (scoped ref R r) => { } }[0];

`// 2`

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:10959 in f00f7d8. [](commit_id = f00f7d83a524f9336843ca496436223e89af9897, deletion_comment = False)

**Comment by cston** [comment]
>         [Fact, WorkItem(49315, "https://github.com/dotnet/roslyn/issues/49315")]

Consider changing to:
```
[Theory]
[InlineData(LanguageVersion.CSharp10)]
[InlineData(LanguageVersion.CSharp11)]
```

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:10955 in 8ec25d7. [](commit_id = 8ec25d77da87cee1e7974f6678170ac9819e290e, deletion_comment = False)

**Comment by cston** [comment]
>             var expectedDiagnostics = new[]

Can be removed.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13607 in 8ec25d7. [](commit_id = 8ec25d77da87cee1e7974f6678170ac9819e290e, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 1]
```diff
@@ -13615,10 +13615,28 @@ public static void M(ref S1 s1)
...
+                //         M2(ref s1, $"{s}" + $"");
+                Diagnostic(ErrorCode.ERR_CallArgMixing, @"M2(ref s1, $""{s}"" + $"""")").WithArguments("CustomHandler.M2(ref S1, ref CustomHandler)", "handler").WithLocation(17, 9),
+                // (17,16): error CS8156: An expression cannot be used in this context because it may not be passed or returned by reference
+                //         M2(ref s1, $"{s}" + $"");
+                Diagnostic(ErrorCode.ERR_RefReturnLvalueExpected, "s1").WithLocation(17, 16),
+                // (17,20): error CS8347: Cannot use a result of 'CustomHandler.CustomHandler(int, int, ref S1)' in this context because it may expose variables referenced by parameter 's1' outside of their declaration scope
+                //         M2(ref s1, $"{s}" + $"");
+                Diagnostic(ErrorCode.ERR_EscapeCall, @"$""{s}"" + $""""").WithArguments("CustomHandler.CustomHandler(int, int, ref S1)", "s1").WithLocation(17, 20),
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13636]: >ErrorCode.ERR_EscapeCall [](http://example.com/codeflow?start=27&length=24)

What is being captured that leads to this error?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 2]
```diff
@@ -1394,18 +1385,24 @@ static S1 MayWrap(ref Span<int> arg)
...
-                Diagnostic(ErrorCode.ERR_CallArgMixing, "MayAssign(ref inner, ref rOuter)").WithArguments("Program.MayAssign(ref System.Span<int>, ref Program.S1)", "arg1").WithLocation(23, 13)
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "MayAssign(ref inner, ref rOuter)").WithArguments("Program.MayAssign(ref System.Span<int>, ref Program.S1)", "arg1").WithLocation(23, 13),
+                // (28,32): error CS9077: Cannot return a parameter by reference 'arg1' through a ref parameter; it can only be returned in a return statement
+                //             arg2 = MayWrap(ref arg1);
+                Diagnostic(ErrorCode.ERR_RefReturnOnlyParameter, "arg1").WithArguments("arg1").WithLocation(28, 32),
+                // (28,20): error CS8347: Cannot use a result of 'Program.MayWrap(ref Span<int>)' in this context because it may expose variables referenced by parameter 'arg' outside of their declaration scope
+                //             arg2 = MayWrap(ref arg1);
+                Diagnostic(ErrorCode.ERR_EscapeCall, "MayWrap(ref arg1)").WithArguments("Program.MayWrap(ref System.Span<int>)", "arg").WithLocation(28, 20)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:1405]: Did `-langversion:10` behavior change?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 3]
```diff
@@ -1618,7 +1618,7 @@ static S1 MayWrap(in Span<int> arg)
         {
             get
             {
-                // should be an error, arg1 is not ref-returnable, 'this' is val-returnable
+                // error: ref-safe-to-escape of arg1 is return-noly
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:1621]: >noly [](http://example.com/codeflow?start=63&length=4)

Typo.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 4]
```diff
@@ -1653,7 +1653,19 @@ static S1 MayWrap(in Span<int> arg)
                 Diagnostic(ErrorCode.ERR_EscapeVariable, "inner").WithArguments("inner").WithLocation(27, 29),
                 // (27,22): error CS8350: This combination of arguments to 'Program.S1.this[in Span<int>]' is disallowed because it may expose variables referenced by parameter 'arg1' outside of their declaration scope
                 //         int dummy4 = rOuter[inner];
-                Diagnostic(ErrorCode.ERR_CallArgMixing, "rOuter[inner]").WithArguments("Program.S1.this[in System.Span<int>]", "arg1").WithLocation(27, 22)
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "rOuter[inner]").WithArguments("Program.S1.this[in System.Span<int>]", "arg1").WithLocation(27, 22),
+                // (44,32): error CS9077: Cannot return a parameter by reference 'arg1' through a ref parameter; it can only be returned in a return statement
+                //                 this = MayWrap(arg1);
+                Diagnostic(ErrorCode.ERR_RefReturnOnlyParameter, "arg1").WithArguments("arg1").WithLocation(44, 32),
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:1659]: >ErrorCode.ERR_RefReturnOnlyParameter [](http://example.com/codeflow?start=27&length=36)

Did `-langversion:10` behavior change?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 5]
```diff
@@ -1726,7 +1738,13 @@ static S1 MayWrap(ref Span<int> arg)
                 Diagnostic(ErrorCode.ERR_EscapeVariable, "inner").WithArguments("inner").WithLocation(29, 31),
                 // (29,13): error CS8350: This combination of arguments to 'Program.D2.Invoke(ref Span<int>, ref Program.S1)' is disallowed because it may expose variables referenced by parameter 'arg1' outside of their declaration scope
                 //             MayAssignDel2(ref inner, ref rOuter);
-                Diagnostic(ErrorCode.ERR_CallArgMixing, "MayAssignDel2(ref inner, ref rOuter)").WithArguments("Program.D2.Invoke(ref System.Span<int>, ref Program.S1)", "arg1").WithLocation(29, 13)
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "MayAssignDel2(ref inner, ref rOuter)").WithArguments("Program.D2.Invoke(ref System.Span<int>, ref Program.S1)", "arg1").WithLocation(29, 13),
+                // (34,32): error CS9077: Cannot return a parameter by reference 'arg1' through a ref parameter; it can only be returned in a return statement
+                //             arg2 = MayWrap(ref arg1);
+                Diagnostic(ErrorCode.ERR_RefReturnOnlyParameter, "arg1").WithArguments("arg1").WithLocation(34, 32),
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:1744]: >ERR_RefReturnOnlyParameter [](http://example.com/codeflow?start=37&length=26)

Did `-langversion:10` behavior change?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 6]
```diff
@@ -2255,7 +2273,13 @@ static S1 MayWrap(in Span<int> arg)
         }
     }
 ";
-            CreateCompilationWithMscorlibAndSpan(text, parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion)).VerifyDiagnostics();
+            CreateCompilationWithMscorlibAndSpan(text, parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion)).VerifyDiagnostics(
+                // (25,20): error CS8347: Cannot use a result of 'Program.MayWrap(in Span<int>)' in this context because it may expose variables referenced by parameter 'arg' outside of their declaration scope
+                //             arg1 = MayWrap(arg2);
+                Diagnostic(ErrorCode.ERR_EscapeCall, "MayWrap(arg2)").WithArguments("Program.MayWrap(in System.Span<int>)", "arg").WithLocation(25, 20),
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:2279]: >ErrorCode.ERR_EscapeCall [](http://example.com/codeflow?start=27&length=24)

Did `-langversion:10` behavior change?

**Comment by cston** [comment]
>     R<string> I<string>.F4(ref R<string> r) => default;

`// 2`, etc.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:10586 in edc9150. [](commit_id = edc91506f7e7c7c1ca66819048d5c7dd694271cc, deletion_comment = False)

**Comment by cston** [comment]
>                 // (9,22): error CS9063: UnscopedRefAttribute cannot be applied to this item because it is unscoped by default.

`ErrorCode.ERR_UnscopedScoped` seems clearer here.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:15948 in edc9150. [](commit_id = edc91506f7e7c7c1ca66819048d5c7dd694271cc, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13626]: I believe what's happening here is that the constructor invocation is being analyzed like its signature is:

`static CustomHandler .ctor(int, int, ref S1)`

when its signature is really:

`static void .ctor(out CustomHandler, int, int, ref S1)`

I think this error is not strictly necessary for safety and would be addressed by #64155.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:1405]: No, I messed up here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:1673]: No, I messed up here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:1758]: No, I messed up here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:2293]: No, I messed up here.

**Comment by RikkiGibson** [comment]
> > ```
> >             // (9,22): error CS9063: UnscopedRefAttribute cannot be applied to this item because it is unscoped by default.
> > ```
> 
> `ErrorCode.ERR_UnscopedScoped` seems clearer here.
> 
> Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:15948 in [edc9150](https://github.com/dotnet/roslyn/commit/edc91506f7e7c7c1ca66819048d5c7dd694271cc). [](commit_id = [edc9150](https://github.com/dotnet/roslyn/commit/edc91506f7e7c7c1ca66819048d5c7dd694271cc), deletion_comment = False)

I don't disagree, but I also think the current diagnostic is sufficiently clear/actionable, so I won't plan to change it in this PR.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs]: Thanks for the explanation. Perhaps add a comment to the test with a link to [https://github.com/dotnet/roslyn/issues/64155](https://github.com/dotnet/roslyn/issues/64155).

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs (hunk 6 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:2353]: It looks like the `-langversion:10` behavior has still changed. Is the test currently failing?

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs]: Adding a note to address in a follow-up.

**Pull request merged**